### PR TITLE
Add linking docs

### PR DESCRIPTION
**Pull request by mooreds**
This addresses https://github.com/FusionAuth/fusionauth-site/issues/864

### CODE REVIEW

**Comment by mooreds** [comment]
@jobannon @robotdan Merging this today

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 1]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+=== Anonymously Link
+
+Richard is logging into NewApp. He doesn't have an account in FusionAuth.
+
+* Richard clicks on the 'Login With Hooli' button on the login screen. 
+* They log in with `richard@hooli.com`, his account at Hooli.
+* He is redirected to FusionAuth.
+* There is an account created in FusionAuth with no username or email address. It is not a full account.
+* Richard can interact with NewApp (a JWT is issued, etc), but is not to use FusionAuth workflows like 'forgot password'.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:220]: >is not to use FusionAuth workflows like 'forgot password'.

But cannot use FusionAuth workflow like ....

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 2]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+==== Anonymously Link, No Email Or Username Returned
+
+Richard is logging into NewApp. He doesn't have an account in FusionAuth. The identity provider is the Hooli XYZ server. This does not return a username or password.
+
+* Richard clicks on the 'Login With Hooli XYZ' button on the login screen. 
+* He logs in with `richard@hoolixyz.com`.
+* He is redirected to FusionAuth.
+* There is an account created in FusionAuth with no username or email address. It is not a full account.
+* Richard can interact with NewApp (a JWT is issued, etc), but is not to use FusionAuth workflows like 'forgot password'.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:232]: But cannot use FusionAuth workflow like ....

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 3]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+Richard is logging into NewApp. He doesn't have an account in FusionAuth.
+
+* Richard clicks on the 'Login With Hooli' button on the login screen. 
+* He logs in with the `richard@piedpiper.com` account. 
+* He is redirected to FusionAuth.
+* A new account is created in FusionAuth with the email `richard@piedpiper.com`. 
+* The new FusionAuth account with the email `richard@piedpiper.com` is linked to the Hooli `richard@piedpiper.com` account.
+
+=== Link By Username For Existing User
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:172]: I am not sure if this reads correctly.  

Link By Username For Existing User --> do not create a user if they don't exist

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 4]
```diff
@@ -8,9 +8,22 @@ description: FusionAuth Identity Providers
...
+* <<Identity Providers and Applications>>
+* <<Overrides>>
+* <<Hints>>
+* <<Linking Strategies>>
+* <<Linking and Create Registration>>
+* <<Linking Examples>>
+** <<Link By Email For Existing User>>
+** <<Link On Email >>
+** <<Link By Username For Existing User>>
+** <<Link On Username >>
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:22]:
```suggestion
** <<Link On Username>>
```

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 5]
```diff
@@ -8,9 +8,22 @@ description: FusionAuth Identity Providers
...
+Identity Providers allow you to enable third-party login in FusionAuth.
+
+* <<Identity Providers and Applications>>
+* <<Overrides>>
+* <<Hints>>
+* <<Linking Strategies>>
+* <<Linking and Create Registration>>
+* <<Linking Examples>>
+** <<Link By Email For Existing User>>
+** <<Link On Email >>
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:20]:
```suggestion
** <<Link On Email>>
```

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 6]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+* `Anonymously Link` - Always create a link based upon the unique Id returned by the identity provider. A username or email is not required and will not be used to link the user. A reconcile lambda will not be used in this configuration.
+* `Link On Email` - Link to an existing user based upon email. A user will be created with the email returned by the identity provider if one does not already exist.
+* `Link On Email. Do not create the user if they do not exist` - Only link to an existing user based upon email. A user will not be created if one does not already exist with email returned by the identity provider.
+* `Link On Username` - Link to an existing user based upon username. A user will be created with the username returned by the identity provider if one does not already exist.
+* `Link On Username. Do not create the user if they do not exist` - Only link to an existing user based upon username. A user will not be created if one does not already exist with the username returned by the identity provider.
+
+Here's a table illustrating the alternatives:
+
+[cols="1,1,1,3"]
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:115]: [frame=all cols="1,1,1,3"]

makes the table read better

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 7]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+|===
+| Strategy | User must exist in FusionAuth | User linked on | Use when the identity provider
+
+| Create a Pending Link | Yes | Yes, user must login to existing account | Has a different email or username than an existing FusionAuth identity and users know enough to link them.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:119]: I don't think that they have to exist within FusinAuth.  Our OAuth flow presents this flow
Screen Shot 2021-10-04 at 4 27 23 PM

So you can create a user

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 8]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+| Anonymously Link | No | IdP id | Exposes neither username nor email.
+| Link On Email | No | Email address | Has the user's email and users that do not exist in the identity provider can have access.
+| Link On Email. Do not create the user if they do not exist | Yes | Email address | Has the user's email and you don't want users that do not exist in the identity provider to have access.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:122]:
```suggestion
| Link On Email. Do not create the user if they do not exist | Yes | Email address | Has the user's email and you don't want users that do not exist in the identity provider to have access.
```

I don't read it this way.  I see this as, joshua@google.com is a google user.  If they want to use their federated identity to login to my app, I better have explicitly created them in FusionAuth first.

image

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 9]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+| Link On Username | No | Username | Has the user's username and users that do not exist in the identity provider can have access.
+| Link On Username. Do not create the user if they do not exist | Yes | Username | Has the user's username and you don't want users that do not exist in the identity provider to have access.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:124]: See above for thoughts

Comment (jobannon) [reply site/docs/v1/tech/identity-providers/index.adoc:124]: > you don't want users that do not exist in the identity provider to have access.

in other words, I don't think that this statement is accurate.  Or I am really confused (either way 🙂 )

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 10]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+The [field]#Linking strategy# and [field]#Create registration# configurations are related to each other, but distinct. 
+The [field]#Linking strategy# controls how a User is created in FusionAuth based on information returned from the remote identity provider. [field]#Create registration# controls if the User created in FusionAuth is registered for a given Application.
+
+== Linking Examples
+
+Here are some walkthroughs of linking scenarios. A user logs in to a remote Identity Provider when the FusionAuth hosted login pages are being used. Similar behavior is available via the link:/docs/v1/tech/apis/identity-providers/[Identity Provider API]. 
+
+It doesn't matter if the Identity Provider is a social provider like Facebook or an enterprise provider like an OIDC or SAML compatible identity server. 
+
+=== Link By Email For Existing User
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:141]: I might have this say - 

> Link By Email.  Do not create user if they do not exist in FusionAuth

**Review by jobannon — COMMENTED** [review]
A few thoughts.

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 11]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+=== Link By Username For Existing User
+
+Richard is logging into NewApp. He doesn't have an account in FusionAuth.
+
+* Richard clicks on the 'Login With Hooli' button on the login screen. 
+* He logs in with the `richard` account. 
+* He is redirected to FusionAuth but sees an error. No access is allowed.
+
+==== Link By Username For Existing User With an Existing Account
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:180]: Should this be `===`

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 12]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+* Richard clicks on the 'Login With Hooli' button on the login screen. 
+* He logs in with the `richard@piedpiper.com` account. 
+* He is redirected to FusionAuth but sees an error. No access is allowed.
+
+Here's an example of the error page:
+
+image::identity-providers/idp-linking-user-must-exist.png[Error when a user must exist for successful linking.,width=1200,role=bottom-cropped]
+
+==== Link By Email For Existing User With an Existing Account
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:153]: Should this be ===

File: site/docs/v1/tech/identity-providers/index.adoc [hunk 13]
```diff
@@ -76,4 +95,143 @@ An email address or domain may be provided in the `login_hint` request parameter
...
+
+* Richard clicks on the 'Login With Hooli' button on the login screen. 
+* They log in with `richard@hooli.com`, his account at Hooli.
+* He is redirected to FusionAuth.
+* There is an account created in FusionAuth with no username or email address. It is not a full account.
+* Richard can interact with NewApp (a JWT is issued, etc), but is not to use FusionAuth workflows like 'forgot password'.
+
+image::identity-providers/idp-linking-anonymous-account.png[Admin view of a user who has linked anonymously.,width=1200,role=bottom-cropped]
+
+==== Anonymously Link, No Email Or Username Returned
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/index.adoc:224]: Should this be ===

**Review by jobannon — CHANGES_REQUESTED** [review]
Add a few more comments on levels for each sub header (ie - `====` vs `===`)

I think that each one of these flows should be reviewed as well.

```
** <<Link By Email For Existing User>>
** <<Link On Email>>
** <<Link By Username For Existing User>>
** <<Link On Username>>
** <<Pending Link>>
** <<Anonymously Link>
```

As I think the actual behavior may be different than what is described.

File: site/docs/v1/tech/identity-providers/index.adoc (hunk 8 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/index.adoc:122]: You are 100% correct. Logic error on my part.

File: site/docs/v1/tech/identity-providers/index.adoc (hunk 13 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/index.adoc:224]: Same thing, this is a subcase.

File: site/docs/v1/tech/identity-providers/index.adoc (hunk 7 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/index.adoc]: It was because self service registration was enabled for the application.

**Pull request merged**